### PR TITLE
make bootstrap use wget if available

### PR DESCRIPTION
**Pull request by lolbinarycat**
this improves reliability when downloading over unreliable connections.

without this change, it was basically impossible for me to run the bootstrap tests.

### CODE REVIEW

**Comment by lolbinarycat** [comment]
whoops, didn't mean to touch that submodule

**Review by Kobzol — COMMENTED** [review]
Hi, thanks for the PR! Just out of curiosity, if you're having issues with network connection, wouldn't `wget -c` be useful for you? With that flag wget could actually resume interrupted downloads.

**Comment by lolbinarycat** [comment]
doesn't `wget -c` only matter if the download previously failed? and wget should automatically retry failures a few times before the whole command fails, right?

just swapping in wget seemed to be enough in my experience.

**Comment by Kobzol** [comment]
Yes, it only matters when you want to resume, I thought that is the situation that you want to improve. So if I understand it correctly, when curl retries a download, it starts from the beginning, but wget restart it where it ended?

Overall, I'm fine with using wget, although I think that we should try to copy some of the options from the curl invocation. For example, there is no connection timeout in wget by default, but for curl we set it to 30s. We should try to get these options for wget and curl to be similar-ish.

**Comment by lolbinarycat** [comment]
> Yes, it only matters when you want to resume, I thought that is the situation that you want to improve.

as i understand it, wget will automatically reconnect and resume by default, and you only need to use `-c` if you want to *manually* continue after that has failed (or after you did something like a ^C)

**Comment by Kobzol** [comment]
I see. Yeah, that makes sense, ok. Could you please try to configure the wget invocations to have more similar options as the curl invocations, particularly around (connection) timeouts, retry counts and similar?

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -100,22 +101,45 @@ def _download(path, url, probably_big, verbose, exception):
...
+                #   but raise `CalledProcessError` or `OSError` instead
+                require(["curl", "--version"], exception=platform_is_win32())
+                run(["curl", option,
+                     "-L", # Follow redirect.
+                     # timeout if speed is < 10 bytes/sec for > 30 seconds
+                     "-y", "30", "-Y", "10",
+                     # timeout if cannot connect within 30 seconds
+                     "--connect-timeout", "30",
+                     "-o", path,
+                     # -S: show errors, even if -s is specified
```
Comment (Kobzol) [inline src/bootstrap/bootstrap.py:136]: Thanks for adding these comments! :)

**Review by Kobzol — COMMENTED** [review]
Thank you! Could you now please squash the commits? There are a lot of tmp/WIP commits in there now.

@onur-ozkan Any concerns against using `wget` as the default downloader if it is available? I'd personally merge this and if people encounter some issues with wget, we can e.g. hide it behind an opt-in env. var. flag.

**Comment by bjorn3** [comment]
Would passing `--continue-at -` to curl work too? According to the man page that is supposed to restart the download from the last downloaded byte when retrying.

**Comment by onur-ozkan** [comment]
> Would passing `--continue-at -` to curl work too? According to the man page that is supposed to restart the download from the last downloaded byte when retrying.

Some of our runners are using very old versions of curl and `--continue-at` is too recent feature for us to use. There are various useful other flags available, but we cannot use them due to this problem.

> @onur-ozkan Any concerns against using wget as the default downloader if it is available? I'd personally merge this and if people encounter some issues with wget, we can e.g. hide it behind an opt-in env. var. flag.

If we use recent features of `wget`, runners will probably fail due to having too old `wget` version.

**Comment by Kobzol** [comment]
I suppose that the easiest way to check is just to try to merge this.

Could you please squash the commits? Thanks.

**Comment by Kobzol** [comment]
Okay, let's try.

@bors r+ rollup=never

**Comment by Kobzol** [comment]
Hmm, unfortunate. I wonder if we could just test if `curl` supports the resume flag, and use it in that case, instead of introducing a full additional way of downloading. Or, if it's annoying or impossible to check properly, just add an environment variable to opt into this, which could then be used by users whose network connection is not stable.

**Comment by onur-ozkan** [comment]
> I wonder if we could just test if curl supports the resume flag

That flag is quite recent feature for `curl` (or at least, for the one we use in our CI runners). I am quite sure that trying that flag will fail too.

> just add an environment variable to opt into this, which could then be used by users whose network connection is not stable.

Makes sense.

**Comment by Kobzol** [comment]
I didn't mean trying the flag for CI, but for local usage (on CI the flag would fail), assuming that recent distros already have new enough curl.

**Comment by onur-ozkan** [comment]
Actually, [it seems](https://github.com/curl/curl/blob/4111d10803cc3b81265f12cdf147750ca3031b60/docs/options-in-versions#L31) that the flag was introduced in early versions of curl. Even our CI runners use newer versions than that (as far as I remember, the major version was either 6 or 7 but definitely not lower or higher). We could try using it unconditionally.

**Comment by Kobzol** [comment]
@lolbinarycat If you're able to replicate your network issues, could you please try if the `--continue-at` curl flag resolves your issue?

**Comment by lolbinarycat** [comment]
@Kobzol --continue-at works, but only in conjunction with `--retry-all-errors`, otherwise a broken pipe or connection terminated by peer error will be incorrectly treated as a non-transient error.

**Comment by Kobzol** [comment]
Oh, shoot. That is sadly a quite [new flag](https://github.com/rust-lang/rust/issues/110178#issuecomment-1506857729).

**Comment by onur-ozkan** [comment]
We can check the curl version and enable it only if it's newer than 7.70.0 (as it was [introduced with 7.71.0](https://github.com/curl/curl/blob/4111d10803cc3b81265f12cdf147750ca3031b60/docs/options-in-versions#L211)). Then, we can close https://github.com/rust-lang/rust/issues/110178 as well.

**Comment by onur-ozkan** [comment]
Closing this one as we are landing #129134.

**Comment by lolbinarycat** [comment]
i believe this actually improves the documentation of what the existing curl flags do, and those comments never got added to the accepted PR

should i add those comments in a seperate PR? or is a whole pr for a couple of comments too frivolous?

**Comment by onur-ozkan** [comment]
Making a separate PR sounds good to me.

**Pull request closed without merging**